### DISCUSSION TITLE
Repeater using TCP with same domain ids

### DISCUSSION THREAD

**Discussion by CselecT** [eProsima/DDS-Router#374] [Q&A]
Hello, first of all thank you for the detailed examples.
I am trying to create a router network that will connect ros2 nodes that are working in different networks. The basic structure is like [here](https://github.com/eProsima/DDS-Router/blob/main/ddsrouter_test/compose/test_cases/repeater_tcp/compose.yml), however I need to be able to set the same domain ids for the sub and pub edges in different networks. I am aware that we can do this with the repeater using udp, however I need to use tcp. I also know that a local discovery server can help, however I could not integrate a local discovery server to the repeater example. I tried starting the discovery server in router edges, but I could not listen the talker from the other network.
To sum up, I want to create [this](https://github.com/eProsima/DDS-Router/blob/main/ddsrouter_test/compose/test_cases/repeater_tcp/compose.yml) example but with using the same domain ids for the pub_edge_1 and sub_edge_2, using tcp. I would be grateful if you can provide an example or direct me to a possible solution. Thanks!

**Comment by jparisu** [comment]
Hi @CselecT 
As this is a question regarding a particular use case rather than an issue, I transfer it to discussions.

**Comment by jparisu** [comment]
Regarding your use case, it would be interesting to check the following [example](<https://eprosima-dds-router.readthedocs.io/en/latest/rst/examples/repeater_example.html>) in the documentation.
I think this example simulates exactly your use case:

- Two edges using the same domain with this [configuration file](<https://github.com/eProsima/DDS-Router/blob/main/docs/resources/examples/repeater_client.yaml>)
- One repeater server connecting both using TCP with this [configuration file](<https://github.com/eProsima/DDS-Router/blob/main/docs/resources/examples/repeater_server.yaml>)

**Reply by CselecT** [reply]
Hello, thank you for your answer.
Actually this was what I tried initially, however I realized that messages sent from one edge to another somehow keeps propagating and I keep receiving the same message from the listener repeatedly. In this screenshot, the first terminal is from the simple participant of the one edge, I ran demo_nodes_cpp talker and in the second terminal, which is from the simple participant of the other edge, I ran the listener. 
![Screenshot from 2023-05-03 10-03-46](https://user-images.githubusercontent.com/60482850/235851769-6d4be012-377e-4d4e-8032-214e39832b13.png)

I also came across with this propagation with a single router but I solved it by using a discovery server inside the router and connected participants to that server. Currently I can build the network that I want with not using routers but only fastdds discovery, but I would love to be able to do that with routers as well. So I would be grateful if anyone solves this message propagation would share the solution here. I can also share my fastdds discovery solution in private, since I am not sure if this is the place for that. 

Thank you for your time!

**Reply by jparisu** [reply]
Ok, so this is a different issue.
The DDS Router currently has a hard limitation: **it cannot connect networks that can be connected without a DDS Router.**
Why is that? Because the DDS Router creates internally DDS entities that receives everything that is happening in one side, and forward all the information that come from the other side. Thus, if the DDS entities in one side are capable of communicating with the entities in the other side, they will create a feedback loop that collapse the application.
You can see that this limitation is documented in the following [page](<https://eprosima-dds-router.readthedocs.io/en/latest/rst/user_manual/configuration.html#participant-configuration>).

Thus, the 2 sides of the DDS Router must be inaccessible to each other. You must choose one of these ways to accomplish your scenario:

- The participants in both sides are in different domains
- The participants in both sides use different discovery servers.
- You can try the branch that is open with a feature that fix this: https://github.com/eProsima/DDS-Router/pull/284. However, this feature is not ready to be released yet, and that branch is not being updated, so you may encounter other issues.
